Add initial module for managing AWS VPC network ACLs

herbyg-axial opened the pull request:
This module provides support for creating, associating, disassociating and modifying AWS VPC network access control lists.  The module allows VPCs and subnets to be specified by their ID, CIDR block or name tag.

If a ruleset and a network ACL identifier is provided, the network ACL is located by that identifier and modified to match the ruleset.  If one can't be located, then a new ACL will be created as per the specified ruleset.  If subnets are specified, they will be associated (or disassociated) as per given configuration.

herbygillot committed:
Add initial module for managing AWS VPC network ACLs

Signed-off-by: Herby Gillot <herby.gillot@gmail.com>

rwe commented:
I like the CIDR/ID/tag functionality here. I'm going to see if I can do something similar for #322 for subnets/route-tables.

herbygillot commented:
Cool.  As I mentioned in #322, I'll make some changes so that the parameters for this module are consistent with the other VPC modules.

herbygillot committed:
ec2_vpc_nacl: clean up documentation and workflow, add module parameter
alises for consistency with other VPC modules, add "max" alias for
specifying the maximum port number

Signed-off-by: Herby Gillot <herby.gillot@gmail.com>

herbygillot commented:
Changes made: added `name`, `rules`, `rules_egress`, `resource_tags` and `vpc_id` parameter aliases for consistency with other VPC modules, among other things.

kalefranz commented:
This is nice code. Too bad all of the AWS ansible code wasn't written in this style.

herbygillot commented:
I still feel that it's pretty terrible, but I appreciate your comment, @kalefranz

phraktle commented:
+1 for merging this (and #322)

bcoca commented on cloud/amazon/ec2_vpc_nacl.py:
```diff
@@ -0,0 +1,1472 @@
+#!/usr/bin/env python
```
shebang should always be /usr/bin/python

bcoca commented on cloud/amazon/ec2_vpc_nacl.py:
```diff
@@ -0,0 +1,1472 @@
+    description:
+      - An identifier (ACL ID, CIDR, tagged name) identifying a network ACL.
+      - If this parameter isn't specified, then a ruleset, if present, will be used to match a NACL with the same rules.
+    required: false
```
when requierd is false, a default should be specified

bcoca commented on cloud/amazon/ec2_vpc_nacl.py:
```diff
@@ -0,0 +1,1472 @@
+description:
+    - This module can create, modify or remove VPC network ACLs.
+    - Can additionally associate/disassociate VPC subnets to network ACLs.
+version_added: 1.8
```
change to "2.0"

bcoca commented on cloud/amazon/ec2_vpc_nacl.py:
```diff
@@ -0,0 +1,1472 @@
+    required: false
+  name:
+    description:
+      - Alias, same as 'nacl'
```
required?

bcoca commented on cloud/amazon/ec2_vpc_nacl.py:
```diff
@@ -0,0 +1,1472 @@
+
+
+#  imports ####################################################################
+from boto.exception import EC2ResponseError
```
for required libs use try/except and set a variable to let user know  in a nice way, otherwise it will generate an exception which is always a bug from a  module. you can look at ec2 module as an example on how to do this.

erydo replied:
Yep, that's an oversight here. Thanks

bcoca commented on cloud/amazon/ec2_vpc_nacl.py:
```diff
@@ -0,0 +1,1472 @@
+    module = get_ansible_module()
+    results = ActionResults()
+
+    try:
```
we shy away from generic/catch all exception handling because it just shows the underlying library messages, most are very unhelpful. We try to prep-end a useful message using the specific context of the action the module was trying to engage in, possibly pointing at problems/solutions.

rwe replied:
This is not a catchall exception clause—these exception types have user-directed error messages.

bcoca replied:
point taken, it just seemed that not all exception points were coverd, like the line bellow that executes get_action_sequence

gregdek commented:
Looks to me like some changes still need to be made to comply with @bcoca comments above -- marking this needs_revision for now. (But should be able to merge quickly since that looks like a pretty comprehensive review.)

gregdek commented:
Hi @herbyg-axial -- just a friendly ping here to ask if you intend to make the changes recommended by @bcoca. Thanks!

herbyg-axial commented:
@gregdek Yes, I will update this soon.

gregdek commented:
@herbyg-axial A friendly reminder: this pull request has been marked as needing your action, and it hasn't been updated in the last month. 

If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we'll keep it open pending your changes.

If we receive no indication from you within a week, we will close this pull request.

gregdek commented:
I am (reluctantly) closing this PR for two reasons: 1. inaction, and 2. new modules are now going into ansible-modules-extras. @herbyg-axial, thanks for the PR, and strongly encourage you to resubmit to the extras repo.

mmochan commented:
@herbyg-axial I was wondering if you were planning on resurrecting this module? If not would you mind if take it over?

jonhadfield commented:
I need a module for managing network ACLs so was planning to write one. If someone is already picking this up, I'm happy to help with development and/or testing.

mmochan commented:
@jonhadfield I'll pick this up and let you know if I need help later. I'm hoping to have something crafted in the next week.

mmochan commented:
@jonhadfield I have a working version of the ec2_vpc_nacl module. I will submit this in the next couple days for review once I've refactored it.

mysysadmin-ltd commented:
Hi @mmochan - thank you for picking this up! Let me know if you want any help with this...

mmochan commented:
Hi @mysysadmin-ltd @jonhadfield 

New pull request https://github.com/ansible/ansible-modules-extras/pull/1502

This is still a work in progress with regards to refactoring but is fully  functional. If you feel like testing and providing feedback that would be very much appreciated.

jonhadfield commented:
Will take a look on the weekend. Thanks.

mysysadmin-ltd commented:
Cheers Mike - will give it a go!

On Fri, Jan 15, 2016 at 11:16 AM, Jon Hadfield notifications@github.com
wrote:

> Will take a look on the weekend. Thanks.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/ansible/ansible-modules-core/pull/403#issuecomment-171936295
> .

mysysadmin-ltd commented:
Hi Mike - sorry for the delay.
Have just given it a whirl and it creates and removes an acl no problem.
Cosmetic things: I stumbled on "vpc" as opposed to "vpc_id" (and I suppose "nacl" should be "name").
Will be adding more acls to my playbook and testing - will update here if necessary.
Thank you!

mmochan commented:
Thanks for the feedback David.  I have a couple of other changes that I will be submitting soon and will also look at your suggestions.
